File: app/src/main/java/com/diamond/appcliente/viewmodel/UsuarioViewModel.java
package com.diamond.appcliente.viewmodel;

import android.content.Context;
import android.net.Uri;
import android.util.Log;
import com.diamond.appcliente.api.ApiClient;
import com.diamond.appcliente.api.AuthApiService;
import com.diamond.appcliente.dto.common.ApiResponse;
import com.diamond.appcliente.dto.usuario.UsuarioDto;
import com.diamond.appcliente.dto.usuario.UsuarioResponse;
import com.google.gson.Gson;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UsuarioViewModel {

    public interface ActualizarCallback {
        void onError(String str);

        void onSuccess(String str);
    }

    public interface UsuarioCallback {
        void onError(String str);

        void onSuccess(UsuarioDto usuarioDto);
    }

    public void obtenerMiUsuario(Context context, final UsuarioCallback callback) {
        ((AuthApiService) ApiClient.getRetrofit(context, true).create(AuthApiService.class)).obtenerMiUsuario().enqueue(new Callback<UsuarioResponse>() {
            public void onResponse(Call<UsuarioResponse> call, Response<UsuarioResponse> response) {
                if (!response.isSuccessful() || response.body() == null) {
                    callback.onError("Error al obtener usuario");
                } else {
                    callback.onSuccess(((UsuarioResponse) response.body()).getData());
                }
            }

            public void onFailure(Call<UsuarioResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }

    public void actualizarMiPerfil(Context context, UsuarioDto dtoUsuario, Uri imagenUri, ActualizarCallback callback) {
        UsuarioDto usuarioDto = dtoUsuario;
        Uri uri = imagenUri;
        final ActualizarCallback actualizarCallback = callback;
        Context context2 = context;
        AuthApiService api = (AuthApiService) ApiClient.getRetrofit(context2, true).create(AuthApiService.class);
        if (usuarioDto.getNombre() != null) {
            if (!usuarioDto.getNombre().isEmpty()) {
                if (usuarioDto.getApellido() != null) {
                    if (!usuarioDto.getApellido().isEmpty()) {
                        if (usuarioDto.getCelular() != null) {
                            if (!usuarioDto.getCelular().isEmpty()) {
                                RequestBody requestBody = RequestBody.create(MediaType.parse("application/json"), new Gson().toJson(usuarioDto));
                                Log.d("UsuarioActivity", "Datos del usuario enviados: " + new Gson().toJson(usuarioDto));
                                MultipartBody.Part imagenPart = null;
                                if (uri != null) {
                                    try {
                                        InputStream inputStream = context2.getContentResolver().openInputStream(uri);
                                        File file = new File(context2.getCacheDir(), "user_image.jpg");
                                        OutputStream outputStream = new FileOutputStream(file);
                                        byte[] buffer = new byte[1024];
                                        while (true) {
                                            int read = inputStream.read(buffer);
                                            int length = read;
                                            if (read == -1) {
                                                break;
                                            }
                                            outputStream.write(buffer, 0, length);
                                        }
                                        inputStream.close();
                                        outputStream.close();
                                        imagenPart = MultipartBody.Part.createFormData("imagen", file.getName(), RequestBody.create(MediaType.parse("image/*"), file));
                                        Log.d("UsuarioActivity", "Imagen seleccionada: " + uri.toString());
                                    } catch (IOException e) {
                                        e.printStackTrace();
                                        Log.e("UsuarioActivity", "Error al manejar la imagen", e);
                                    }
                                }
                                api.actualizarMiPerfil(requestBody, imagenPart).enqueue(new Callback<ApiResponse<Object>>() {
                                    public void onResponse(Call<ApiResponse<Object>> call, Response<ApiResponse<Object>> response) {
                                        if (response.isSuccessful()) {
                                            Log.d("UsuarioActivity", "Respuesta exitosa: " + ((ApiResponse) response.body()).toString());
                                            actualizarCallback.onSuccess("Usuario actualizado exitosamente");
                                            return;
                                        }
                                        Log.e("UsuarioActivity", "Error en la respuesta: " + response.code() + " - " + response.message());
                                        try {
                                            Log.e("UsuarioActivity", "Cuerpo de la respuesta de error: " + response.errorBody().string());
                                        } catch (IOException e) {
                                            Log.e("UsuarioActivity", "Error al leer el cuerpo de la respuesta", e);
                                        }
                                        actualizarCallback.onError("Error al actualizar usuario");
                                    }

                                    public void onFailure(Call<ApiResponse<Object>> call, Throwable t) {
                                        Log.e("UsuarioActivity", "Error al actualizar usuario", t);
                                        actualizarCallback.onError(t.getMessage());
                                    }
                                });
                                return;
                            }
                        }
                        actualizarCallback.onError("El campo celular no puede estar vacío");
                        return;
                    }
                }
                actualizarCallback.onError("El campo apellido no puede estar vacío");
                return;
            }
        }
        actualizarCallback.onError("El campo nombre no puede estar vacío");
    }
}

File: app/src/main/java/com/diamond/appcliente/viewmodel/UsuarioViewModel.kt
package com.diamond.appcliente.viewmodel

import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.diamond.appcliente.dto.usuario.UsuarioDto
import com.diamond.appcliente.repository.UsuarioRepository
import com.diamond.barbershop.shared.util.PreferenciasHelper
import com.shared.models.ui.state.UiState
import com.google.gson.Gson
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.File
import java.io.FileOutputStream
import javax.inject.Inject

@HiltViewModel
class UsuarioViewModel @Inject constructor(
    private val usuarioRepository: UsuarioRepository,
    private val preferenciasHelper: PreferenciasHelper,
    @ApplicationContext private val context: Context
) : ViewModel() {

    private val _usuario = MutableStateFlow<UiState<UsuarioDto>>(UiState.Idle)
    val usuario: StateFlow<UiState<UsuarioDto>> = _usuario.asStateFlow()

    private val _actualizarEvento = MutableSharedFlow<String>()
    val actualizarEvento: SharedFlow<String> = _actualizarEvento.asSharedFlow()

    private val _error = MutableSharedFlow<String>()
    val error: SharedFlow<String> = _error.asSharedFlow()

    fun cerrarSesion() {
        context.getSharedPreferences("diamond_prefs", Context.MODE_PRIVATE)
            .edit().putBoolean("welcome_shown", false).apply()
        preferenciasHelper.limpiarPreferencias()
    }

    fun obtenerMiUsuario() {
        viewModelScope.launch {
            _usuario.value = UiState.Loading
            try {
                val result = usuarioRepository.obtenerMiUsuario()
                _usuario.value = if (result != null) UiState.Success(result) else UiState.Empty
            } catch (e: Exception) {
                _usuario.value = UiState.Error(e.message ?: "Error desconocido")
            }
        }
    }

    fun actualizarMiPerfil(nombre: String, apellido: String, email: String, celular: String, imagenUri: Uri?) {
        if (celular.length != 9) {
            viewModelScope.launch { _error.emit("El número de celular debe tener 9 dígitos.") }
            return
        }
        val dto = UsuarioDto().apply {
            this.nombre = nombre
            this.apellido = apellido
            this.email = email
            this.celular = celular
        }
        actualizarMiPerfil(dto, imagenUri)
    }

    private fun actualizarMiPerfil(dtoUsuario: UsuarioDto, imagenUri: Uri?) {
        viewModelScope.launch {
            if (dtoUsuario.nombre.isNullOrEmpty()) { _error.emit("El campo nombre no puede estar vacío"); return@launch }
            if (dtoUsuario.apellido.isNullOrEmpty()) { _error.emit("El campo apellido no puede estar vacío"); return@launch }
            if (dtoUsuario.celular.isNullOrEmpty()) { _error.emit("El campo celular no puede estar vacío"); return@launch }

            val requestBody = RequestBody.create(MediaType.parse("application/json"), Gson().toJson(dtoUsuario))
            Log.d("UsuarioViewModel", "Datos enviados: ${Gson().toJson(dtoUsuario)}")

            val imagenPart: MultipartBody.Part? = imagenUri?.let { uri ->
                try {
                    val inputStream = context.contentResolver.openInputStream(uri)!!
                    val file = File(context.cacheDir, "user_image.jpg")
                    FileOutputStream(file).use { out -> inputStream.copyTo(out) }
                    inputStream.close()
                    Log.d("UsuarioViewModel", "Imagen seleccionada: $uri")
                    MultipartBody.Part.createFormData("imagen", file.name, RequestBody.create(MediaType.parse("image/*"), file))
                } catch (e: Exception) {
                    Log.e("UsuarioViewModel", "Error al manejar la imagen", e)
                    null
                }
            }

            try {
                _actualizarEvento.emit(usuarioRepository.actualizarMiPerfil(requestBody, imagenPart!!))
                obtenerMiUsuario()
            } catch (e: Exception) {
                Log.e("UsuarioViewModel", "Error al actualizar usuario", e)
                _error.emit(e.message ?: "Error desconocido")
            }
        }
    }
}
